great_expectations/data_context/data_context.py

```diff
@@ -927,76 +929,108 @@ def profile_datasource(self,
 
         # KeyError will happen if there are no generators
-        data_asset_name_list = []
+        available_data_asset_name_list = []
         try:
-            datasource = data_asset_names[datasource_name]
+            datasource_data_asset_names_dict = data_asset_names_dict[datasource_name]
         except KeyError:
             # KeyError will happen if there is not datasource
             raise ge_exceptions.ProfilerError(
                 "No datasource {} found.".format(datasource_name))
-        try:
-            for generator_name in datasource.keys():
-                for name in datasource[generator_name]["names"]:
-                    data_asset_name_list.append((generator_name, name[0]))
 
-        except KeyError:
-            data_asset_name_list = []
+        if generator_name is None:
+            # if no generator name is passed as an arg and the datasource has only
+            # one generator with data asset names, use it.
+            # if ambiguous, raise an exception
+            for name in datasource_data_asset_names_dict.keys():
+                if generator_name is not None:
+                    profiling_results = {
+                        'success': False,
+                        'error': {
+                            'code': DataContext.PROFILING_ERROR_CODE_MULTIPLE_GENERATORS_FOUND
+                        }
+                    }
+                    return profiling_results
+
+                if len(datasource_data_asset_names_dict[name]["names"]) > 0:
+                    available_data_asset_name_list = datasource_data_asset_names_dict[name]["names"]
+                    generator_name = name
+
+            if generator_name is None:
+                profiling_results = {
+                    'success': False,
+                    'error': {
+                        'code': DataContext.PROFILING_ERROR_CODE_NO_GENERATOR_FOUND
+                    }
+                }
+                return profiling_results
+        else:
+            # if the generator name is passed as an arg, get this generator's available data asset names
+            try:
+                available_data_asset_name_list = datasource_data_asset_names_dict[generator_name]["names"]
+            except KeyError:
+                raise ge_exceptions.ProfilerError(
+                    "Generator {} not found. Specify generator name")
 
-        if len(data_asset_name_list) == 0:
+        available_data_asset_name_list = sorted(available_data_asset_name_list, key=lambda x: x[0])
+
+        if len(available_data_asset_name_list) == 0:
             raise ge_exceptions.ProfilerError(
-                "No Data Assets found in Datasource {}. Used generators: {}.".format(
+                "No Data Assets found in Datasource {}. Used generator: {}.".format(
                     datasource_name,
-                    list(datasource.keys()))
+                    generator_name)
```

Comment:
This all makes sense after considerable study. However I don't see testing for all the end states.

Reply:
You are right! Adding tests in a separate PR.